File: python/1337.py
# Link: https://leetcode.com/problems/the-k-weakest-rows-in-a-matrix/description/
# Runtime: 98 ms
# Memory: 16.8 MB

class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        frequency_and_numbers_of_rows = [(sum(value), index) for index, value in enumerate(mat)]
        frequency_and_numbers_of_rows.sort(key=lambda x: (x[0], x[1]))

        return [value[1] for value in frequency_and_numbers_of_rows[:k]]

File: c++/1337.cpp
// Link: https://leetcode.com/problems/the-k-weakest-rows-in-a-matrix/description/
// Runtime: 9 ms
// Memory: 11 MB

class Solution {
public:
  std::vector<int> kWeakestRows(const std::vector<vector<int>> &mat, int k) {
    std::map<int, std::vector<int>> frequencyAndNumbersOfRows;
    for (int i = 0; i < static_cast<int>(mat.size()); i++)
      frequencyAndNumbersOfRows[std::reduce(mat[i].begin(), mat[i].end())]
          .push_back(i);

    int numbersFound = 0;
    std::vector<int> weakestRows;
    for (const auto &[key, value] : frequencyAndNumbersOfRows) {
      int i = 0;
      while ((numbersFound < k) && (i < static_cast<int>(value.size()))) {
        weakestRows.push_back(value[i]);
        numbersFound++;
        i++;
      }
    }

    return weakestRows;
  }
};
